enhancements/single-node-production-edge-cluster-profile.md

```diff
@@ -0,0 +1,380 @@
+this DU component needs to be deployed very close to the radio antenna for
+which it is responsible. In practice, this can mean running this
+component on anything from a single server at the base of a remote cell
+tower or in a datacenter-like environment serving several base stations.
+
+A hypothetical DU example is an unusually resource-intensive workload,
+requiring 20 dedicated cores, 24 GiB of RAM consumed as huge pages,
+multiple SR-IOV NICs carrying several Gbps of traffic each, and
+specialized accelerator devices. The node hosting this workload must
+run a realtime
+kernel, be carefully tuned to ensure low-latency requirements
+can be met, and be configured to support features like Precision Timing
+Protocol (PTP).
+
+One crucial detail of this use case is the "cloud" hosting this workload
+is expected to be "autonomous" such that it can continue operating with
+its existing configuration and running the existing workload, even when
+any centralized management functionality is unavailable.
+
+### Goals
+
+* This enhancement describes an approach for deploying OpenShift in
+  single-node configurations for production use in environments with
+  "reasonably significant" memory, storage, and compute resources.
+* Clusters built using the `single-node-production-edge` profile
+  should pass most Kubernetes and OpenShift conformance end-to-end
+  tests. Any tests that must be skipped due to differences from a
+  multi-node deployment will be documented.
+
+### Non-Goals
+
+* This enhancement does not address single-node deployments in
+  highly-constrained environments such as Internet-of-things devices
+  or personal computers.
+* This enhancement does not address "developer" use cases. See the
+  [single-node-developer-profile](https://github.com/openshift/enhancements/pull/302)
+  enhancement.
+* This enhancement does not address high-availability for single-node
+  deployments.
+* This enhancement does not address in-place upgrades for this first
+  iteration. Upgrades will initially only be achieved by redeploying
+  the machine and its workload.
+* This enhancement does not attempt to describe a way to "pre-build"
+  deployment images, either generically or customized for a user.
+* This enhancement does not address the removal of the bootstrap VM,
+  although single-node clusters would benefit from that work, which
+  will be described in a separate enhancement.
+
+## Proposal
+
+After the profile is introduced, all teams developing OpenShift
+components will need to consider how their components should be
+configured when deployed and used in the `single-node-production-edge`
+deployments.
+
+Although the environment is assumed to have significant resources, it
+is important to dedicate most of them to end-user workloads, rather
+than cluster control plane or monitoring. Therefore, the cluster
+profile will configure telemetry and logging to forward data, instead
```

Comment:
Need to consider the case where the single node is disconnected from the centralised collection, need to buffer and forward when connection is re-established.

Reply:
I'm pretty skeptical of this.  The entire point of the prometheus stack is to buffer collection.  If you're inventing a new path for this that has to be recreated for the large variety of system data we collect, that's a red flag to me.

Reply:
To be clearer, I'm not convinced the general single edge production node profile has the resource constraints AND the store and forward requirement described here (I can believe telecom edge does, just not *all* production edge).  Can you make a stronger case for this being a general statement for all production edge?

Reply:
I note this because monitoring is a fundamental component of OpenShift.  It provides the loop by which insight into production performance is measured.  Up until this proposal monitoring has been required to exist, and we assume that monitoring is a fundamental part of the platform, with component being self monitored and self managed.  Saying "we will do this someplace else" removes the closed loop within OpenShift, limits how good operators can be at self observation, duplicates a large amount of configuration, and except in very trivial integration scenarios (i just want app metrics, or just want a small subset of node metrics) is going to duplicate a lot of work.

Cluster monitoring talks to at least 6 on node components (node exporter, kubelet, networking, machine config daemon, dns, and things like SRO).  Which of those contain data that central monitoring needs?  I would wager most.  Which of the 23 core components must be stored and forwarded for alerting of platform health?  I would wager most.  So if we end up duplicating outside the platform a significant fraction of the scope what the platform already gathers, then we're wasting engineering effort that would be better spent on efficiency.

Reply:
Copying here for context: we've discussed with monitoring before this is an almost infinitely tunable component in terms of CPU, memory, and disk space, and I'd expect to see justifications that tuning is insufficient before we remove it:

* going from two replicas to one is half CPU and memory
* cutting retention in half reduces disk in half (and I believe memory, but that may not be true anymore)
* doubling the scrape interval reduces CPU in half
* cutting out half of the metric series scraped (of which 55% of the default cluster metrics are control plane and likely 25% are just bugs we can trivially fix as excessive cardinality) should halve memory and cpu and disk.

And I'm sure there are some more.  Doing the three CPU ones of those might result in 1/8th the CPU and memory OOTB.  I would be completely supportive of "efficient single node monitoring" (since that benefits everyone), but I'm generally not supportive of "monitoring off".  A single prometheus should be able to run at about 0.1 or 0.2 core, potentially even before some of these tuning.  Have we collected the data and the tuning before we jump to "turn off"?

Reply:
agreeing what @smarterclayton has said so far, and cc'ing @simonpasquier. We have also discussed the above options already with @dhellmann and have an investigational spike on what we can actually achieve.